SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestNetReceive/SpatialTestNetReceive.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialTestNetReceive.h"
+
+#include "SpatialFunctionalTestFlowController.h"
+#include "SpatialGDK/Public/EngineClasses/SpatialNetDriver.h"
+
+#include "Net/UnrealNetwork.h"
+
+/**
+ * This test tests Pre and PostNetReceives, along with their ordering with RepNotifies.
+ *
+ * The test includes 1 Server and 1 Client worker.
+ * The flow is as follows:
+ * - Test:
+ *	- The server spawns a ASpatialTestNetReceiveActor, which has a static subobject. The static subobject is where the properties are
+ *stored. We use an subobject instead of storing properties on the actor as the subobject has significantly fewer replicated properties
+ *inherited from native.
+ *		- The server also sets the values of the replicated int and the replicated OwnerOnly int. The actor's owner is set to the client's
+ *PlayerController.
+ *	- The client then checks callbacks were called in the right order
+ */
+
+static constexpr float StepTimeLimit = 15.0f;
+ASpatialTestNetReceive::ASpatialTestNetReceive()
+	: Super()
+{
+	Author = TEXT("Arthur");
+	Description = TEXT("Test Pre and PostNetReceives, along with their ordering with RepNotifies");
+}
+
+void ASpatialTestNetReceive::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(TEXT("SpatialTestNetReceiveInitialise"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		APlayerController* PlayerController =
+			Cast<APlayerController>(GetFlowController(ESpatialFunctionalTestWorkerType::Client, 1)->GetOwner());
+
+		FActorSpawnParameters Params;
+		Params.Owner = PlayerController;
+		TestActor = GetWorld()->SpawnActor<ASpatialTestNetReceiveActor>(Params);
+		RegisterAutoDestroyActor(TestActor);
+
+		checkf(IsValid(TestActor), TEXT("TestActor must be valid"));
+
+		TestActor->Subobject->TestInt = 5;
+		TestActor->Subobject->OwnerOnlyTestInt = 6;
+
+		FinishStep();
+	});
+
+	AddStep(
+		TEXT("SpatialTestNetReceiveCheckAllOk"), FWorkerDefinition::Client(1), nullptr, nullptr,
+		[this](float Dt) {
+			if (!RequireEqual_Bool(IsValid(TestActor), true, TEXT("TestActor should be valid")))
+			{
+				return;
+			}
+			if (!RequireEqual_Int(IsValid(TestActor->Subobject), true, TEXT("TestActor's subobject should be valid")))
+			{
+				return;
+			}
+
+			RequireEqual_Int(TestActor->Subobject->TestInt, 5, TEXT("TestInt property should be updated"));
```

Comment:
I think everything after the `return` statements can be an Assert instead, since everything should be in order once the test actor is Valid (e.g. we don't have to wait for anything else).

Reply:
yup, good point